Fix documentation around --enable-graphics flag (#224)
canonical/anbox-cloud-docs: explanation/rendering-graphics.md

(exp-rendering-graphics)=

> # Rendering graphical output

Reviewer: We don't use "graphical output" but "visual output" only. Is that planned?

Anbox Cloud can adapt to your rendering needs by adjusting its behaviour depending on the availability of a GPU and the presence of the `--enable-graphics` flag when you launch an instance.

    amc launch --enable-graphics my-application

> The use of `--enable-graphics` flag when you launch an instance decides whether Anbox Cloud renders a visual output.

Reviewer: ```suggestion
The use of the `--enable-graphics` flag when you launch an instance decides whether Anbox Cloud renders a visual output.
```

When this flag is set, if there is a GPU available, Anbox Cloud renders visual output using the GPU. If there is no GPU available, it uses software rendering to render the visual output.

> However, if you do not use the `--enable-graphics` flag when launching an instance, Anbox Cloud does not render a visual output.

Reviewer: ```suggestion
However, if you do not use the `--enable-graphics` flag when launching an instance, Anbox Cloud does not provide any visual output.
```

> Now your environment may or may not include a GPU. If Anbox Cloud detects a GPU device during deployment, it configures the cluster to use the GPU. AMS configures each LXD instance to pass through a GPU device from the host. Currently, all GPUs that are available to a machine are passed to every instance that owns a GPU slot. For NVIDIA GPUs, LXD uses the [NVIDIA container runtime](https://github.com/NVIDIA/nvidia-container-runtime) to make the GPU driver of the host available to the instance.

Reviewer: ```suggestion
Now your environment may or may not include a GPU. If Anbox Cloud detects a GPU device during deployment, it adjusts itself to use the GPU. AMS configures each LXD instance to pass through a GPU device from the host. Currently, all GPUs that are available to a machine are passed to every instance that owns a GPU slot. For NVIDIA GPUs, LXD uses the [NVIDIA container runtime](https://github.com/NVIDIA/nvidia-container-runtime) to make the GPU driver of the host available to the instance.
```

We don't always have a cluster so lets be more generic.

Furthermore `Currently, all GPUs that are available to a machine are passed to every instance that owns a GPU slot.` isn't true. We only pass a single GPU but the entire GPU will be available to the instance. There is no slicing support available.

```{important}

> Anbox Cloud currently does not support GPU provisioning for virtual machines. This feature is planned for a future release.

Reviewer: This can be read in different ways. If the virtual machine has a GPU attached and we run Anbox Cloud on top of it, we can use it. I think what you mean is that if we run a VM instance within AMS with an Anbox image, right?

Author: Yes, the instances that are of type VM and not container - clarified the same.

```

To use container instances with a GPU,

- Configure the number of available GPU slots on the node. This decides how many instances can run on a given node because GPUs have limited sharing capacity amongst multiple instances. See {ref}`sec-gpu-slots` for detailed information.

> - Check the list of supported GPUs ({ref}`sec-supported-gpus`) to see if Anbox Cloud includes a driver for your GPU device. If a GPU driver is available inside the instance, there are no further differences in how to use it in comparison to a regular environment. If no GPU driver is available, you must provide it through an {ref}`addon <exp-addons>`.

Reviewer: We don't include drivers in Anbox Cloud. Maybe

```suggestion
- Check the list of supported GPUs ({ref}`sec-supported-gpus`) to see if Anbox Cloud has support for a driver for your GPU device. If a GPU driver is available inside the instance, there are no further differences in how to use it in comparison to a regular environment. If no GPU driver is available, you must provide it through an {ref}`addon <exp-addons>`.
```

- Launch your instance with the `--enable-graphics` flag.

> If your environment does not have a GPU, Anbox Cloud uses software or headless rendering.

Reviewer: What do you mean by "headless rendering"? Headless means there is no output attached but it is not something which can replace "software rendering".

Author: Ok, so they are not interchangeable. Then, let's talk only about software rendering becauses that's what we use consistenly across our docs.

To use instances without a GPU and still render visual output,

> - Change the {ref}`sec-application-manifest-resources` of the application to not require a GPU. This tells Anbox Cloud that no GPU is available.

Reviewer: ```suggestion
- Change the {ref}`sec-application-manifest-resources` of the application to not require a GPU. This tells Anbox Cloud that no GPU should be used.
```

- Launch your instance with the `--enable-graphics` flag. The presence of this flag when there is no GPU tells Anbox Cloud to use software rendering.

> Since software rendering and video encoding utilise the CPU, you won't be able to run as many instances on a system when compared to running instances with a GPU.

Reviewer: I would drop this as it depends on performance characteristics we don't have under control.
